Adding ability to trigger downstream jobs when the uptream job finishes

vaibhavmagarwal opened the pull request:
This depends on https://github.com/adusca/pulse_actions/pull/8 , not ready for review.

vaibhavmagarwal commented:
r? @adusca 
This is now ready for review. I have tested it out locally with:

``` python
python pulse_actions/worker.py --topic-base=builds_finish
```

and it is correctly tracking builds and working fine. This might also potentially help with try-extender for tracking builds. I think we might want to have a separate worker (low cost) to test this out in production without affecting the existing workers, I will talk to chmanchester about it tomorrow.

vaibhavmagarwal commented on pulse_actions/handlers/builds.py:
```diff
@@ -0,0 +1,38 @@
+        buildername = dictionary['builders'][0]
+        LOG.info("Backfilling jobs after build %s requested by %s has completed." % (build, requester))
+        LOG.info("Buildername to be backfilled: %s" % buildername)
+        trigger_range(buildername, [revision], dry_run=dry_run)
```
We are calling trigger_range, since we just want to trigger the buildername now and not manual_backfill it.

vaibhavmagarwal commented:
After talking with @chmanchester today, we have decided to try out the builds-queuing functionality in the existing worker, and if in future we feel the need we will shift it to another worker.

adusca commented:
How will the extra-properties approach work for builds that are already running? I understand triggering new build jobs with extra-properties on mozci, but if the build is already running I don't think triggering it again just for the extra-properties is a reasonable solution.

adusca commented on pulse_actions/handlers/builds.py:
```diff
@@ -0,0 +1,44 @@
+
+    if dictionary['type'] == 'trigger_missing_jobs_for_revision':
+        LOG.info("Triggering missing jobs after build %s requested by %s has completed." % (build, requester))
+        trigger_missing_jobs_for_revision(repo_name, revision, dry_run=dry_run)
```
I thought the extra-properties approach would give us a list of exactly what to trigger when a build finishes. This way we are trying to trigger ~900 jobs whenever a build we asked for before finishes.

vaibhavmagarwal replied:
We can just call trigger_missing_jobs functions in mozci - it takes ~2-3 minutes to go over all the jobs, and it won't trigger existing jobs (since there is already an instance), just the new jobs. Sending all the buildernames for a build again seemed to complicate things like maintaining a map.

adusca replied:
Calling trigger_range 900 times each time we get a 'trigger_missing_jobs' message has disadvantages.

Even though trigger_range tries its best to avoid multiple triggerings, there are situations where it doesn't work:  a requested job takes some time to show up in buidapi and we had bugs in trigger_range regarding PENDING and FAILED jobs that cause it to trigger unnecessary jobs.

Even if we don't trigger jobs multiple jobs, this is potentially very slow. trigger_missing_jobs can take up to 3 minutes to decide there's no work to be done. This could slow pulse_actions considerably and make it unreliable.

adusca commented on pulse_actions/handlers/builds.py:
```diff
@@ -0,0 +1,44 @@
+        LOG.info("Buildername to be backfilled: %s" % buildername)
+        # We are calling trigger_range, since we just want to trigger the missing buildername, not backfill it.
+        trigger_range(buildername, [revision], dry_run=dry_run)
+    elif dictionary['type'] == 'trigger_all_talos_jobs':
```
In mozci, both for alltalos and --backfill users have requested the ability to only trigger jobs if the build job already exists (especially for alltalos on try). Do you know what sheriffs would prefer to be the default in TH?

vaibhavmagarwal replied:
Asked @jmaher, he said he would like pulse_actions to track builds queuing functionality. I think we should have it, we can make changes if this is not required in future.

adusca replied:
I asked jmaher and he said that for try he doesn't want to trigger talos jobs for builds that were not asked for on the try syntax. We can just special-case try.

vaibhavmagarwal commented:
@adusca you are right that this approach won't work if there is already an existing build running. We won't be triggering a new build with extra properties, instead we won't do anything if a build is running. I had discussed with @chmanchester about this, he told that that one will have to press the button again in this case (for time being).

adusca commented:
I think it would be better if instead of doing different things for different dict types we just triggered the jobs from a list of 'downstream jobs to trigger when this is done' that we receive from extra-properties.

That way we could use this for mozci itself (instead of telling people to wait when there is no build job we create an appropriate extra-properties) and any other use case that needs this functionality.

vaibhavmagarwal commented:
@adusca We need to have different parameters in different functions (like times parameter is used in trigger_all_talos while it is not in other functions). On thinking about this a bit, we don't know beforehand in the trigger_missing_jobs function, about which upstream build is not present. For having a list of buildernames in trigger_missing_jobs, we would need to hack it inside trigger_range function (since that is where we decide whether an upstream builder needs to be triggered), which is not a good way either, since mozci is also used by locally by developers. I think we should just call the functions.

adusca commented:
We can add a times parameter that defaults to 1.

We can modify determine_trigger_objective to receive a list of the buildernames we want to trigger and return a list of (buildername, extra_properties) tuples (extra_properties can be None when we don't need it). That would allow developers who run mozci locally to just run the script once and the test jobs would be automatically triggered when the build job finishes (by pulse actions).

armenzg commented on pulse_actions/handlers/builds.py:
```diff
@@ -0,0 +1,44 @@
+def on_builds_finish_event(data, message, dry_run):
+    # We need to ack the message to remove it from our queue
+    message.ack()
+    text = data['payload']['build']['text']
```
What does 'text' represent? A more meaningful name would help here.

armenzg commented on pulse_actions/handlers/builds.py:
```diff
@@ -0,0 +1,44 @@
+    requester = data['payload']['build']['blame']
+    properties = data['payload']['build']['properties']
+    dictionary = {}
+    for property_list in properties:
```
Would this line work?

> for property_key, property_value in properties:

vaibhavmagarwal replied:
properties is a list here, not a dictionary, so it wouldn't work.

armenzg commented:
I believe this is the PR that chmanchester asked if we could land until we can use TC's abiltiy to run dependent tasks upon an upstream task completing. At that point we could back this code out.

@vaibhavmagarwal is this still the wanted outcome for this PR?

Side note, IIRC I believe this was not solving the case for running builds which is fine for now. Correct me if not.

vaibhavmagarwal commented:
@armenzg yes, this PR works for tracking newly triggered builds (not the existing running ones). When @adusca implements the TC task graph, we can back out this patch. I have addressed the nits. Merge it when you feel it is ok to do so.

adusca commented:
With a small amount of extra work, we get a general scheduling solution which fixes the long-standing issue of upstream jobs on mozci. The current way only works for a subset of use cases and it has disadvantages even for those use cases.

I have the above concerns, but I am not opposed to merging the patch as long as the requests get monitored appropriately for delays and duplicate jobs.

armenzg commented:
@vaibhavmagarwal As Alice points out, we could have a better solution next week and this one might have issues. If you prefer not to wait, please keep an eye on this since we can make pulse_actions unstable or unreliable.

FYI, I hope to have a prototype by EOW and a release by next week.

vaibhavmagarwal commented:
I am fine with waiting, if we are going to be ready in the next 2 weeks with the TC system, we may not deploy this.

armenzg commented:
OK cool. I will update my progess on the issue I just filed earlier.
